subprojects/core/src/main/java/org/gradle/api/internal/initialization/ClassLoaderScope.java

```diff
@@ -60,6 +61,25 @@
      */
     boolean defines(Class<?> clazz);
 
+    /**
+     * Returns the plugin version for the given plugin id, if known to this scope.
+     *
+     * @param pluginId the id of the plugin
+     * @return the version, if known
+     */
+    @Nullable
+    String getPluginVersion(PluginId pluginId);
```

Comment:
It feels a bit weird to have a plugin specific code in ClassLoaderScope. It feels weird since now ClassLoaderScope has a dependency on plugins and it makes it more tightly coupled with plugins.

Maybe we could do something like it is done in VisitableURLClassLoader, with some generic "userData":
`https://github.com/gradle/gradle/blob/5ec3f672ed600a86280be490395d70b7bc634862/subprojects/base-services/src/main/java/org/gradle/internal/classloader/VisitableURLClassLoader.java#L43`
and then we would pass in some "Plugin Scope" object that would have plugin versions info. Since `userData` would be generic we would remove the dependency to plugins from ClassLoaderScope (it would be less safe though).

I guess having some "plugin scope" object passed to a DefaultPluginRequestApplicator would be a big change right?

Reply:
I agree with Anže here. @big-guy any opinion?

Reply:
I have done this.